<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en-AU">

  <head>
  <meta name="DESCRIPTION" content="WebBasedPRogramming.com is your free online resource for learning to program using the .NET Framework, ASP.NET, C#, C, C++, Java, J2EE, SQL, Perl, Web Services, Visual Basic .NET, XML, VB.NET, and many other languages over the internet.">
<meta name="KEYWORDS" content="online, programming, learn, lesson, tutorial, free, J2EE, vb, visual basic, delphi, xml, asp, C++, C, assembly, compiler, program, language, help, C sharp, Java, tutor, perl, sql, script, teach, teaching, courses, internet, .NET Framework, web services, ASP.NET, code, lessons, interactive, software development, programmer, developer">

    <meta http-equiv="content-type" content="application/xhtml+xml; charset=UTF-8" />


    <!-- Navigational metadata for large websites (an accessibility feature): -->


    <link rel="stylesheet" type="text/css" href="sinorca-screen.css" media="screen" title="Sinorca (screen)" />
    <link rel="stylesheet alternative" type="text/css" href="sinorca-screen-alt.css" media="screen" title="Sinorca (alternative)" />
    <link rel="stylesheet" type="text/css" href="sinorca-print.css" media="print" />

    <title>Java Unleashed Second Edition</title>
 
  </head>

  <body>
    <!-- For non-visual user agents: -->
      <div id="top"><a href="#main-copy" class="doNotDisplay doNotPrint">Skip to main content.</a></div>

    <!-- ##### Header ##### -->

    <div id="header">
      <div class="superHeader">
      
      <!-- 
        <div class="left">
        Related sites:
          <a href="./index.html">Link 1</a> |
          <a href="./index.html">Link 2</a></div>
        <div class="right">
          More related sites:
          <a href="./index.html">Link 3</a> |
          <a href="./index.html">Link 4</a> |
          <a href="./index.html">Link 5</a> |
          <a href="./index.html">Link 6</a> |
          <a href="./index.html">Link 7</a> -->
        </div>
      </div>

      <div class="midHeader">
        <h1 class="headerTitle" align="center"><i>
        <font face="Arial Black" size="6">Web Based Programming Tutorials</font></i></h1>
      </div>

      <div class="subHeader"> <center>
        <a href="http://www.webbasedprogramming.com">Homepage</a> |
        <a class="highlight" target="_blank" href="http://www.webbasedprogramming.com/forum/">Forum</a> - Join the forum to 
        discuss anything related to programming! |
        <a href="http://www.webbasedprogramming.com/programming-resources">Programming Resources</a></center>
      </div>
    </div>

    <!-- ##### Side Bar ##### -->

    <div id="side-bar">
      <div>
        <p class="sideBarTitle" align="center">Menu</p>
        <ul>
          <li>&nbsp;</li>
          <li><a href="http://www.webbasedprogramming.com/CGI-Developers-Guide/"> CGI Developer&#39;s Guide</a></li>
          <li><a href="http://www.webbasedprogramming.com/CGI-Manual-of-Style/" >
           CGI Manual of Style</a></li>
          <li><a href="http://www.webbasedprogramming.com/CGI-Programming-Unleashed/" > CGI 
          Programming Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/Creating-Web-Applets-with-Java/" > 
          Creating Web Applets with Java</a></li>
          <li><a href="http://www.webbasedprogramming.com/Developing-Professional-Java-Applets/" > Developing 
          Professional Java Applets</a></li>
          <li><a href="http://www.webbasedprogramming.com/How-to-use-HTML-3.2/" > 
          How to use HTML 3.2</a></li>
          <li><a href="http://www.webbasedprogramming.com/Java-1.1-Unleashed/" > Java 1.1 
          Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/Java-1.2-Unleashed/" > 
          Java 1.2 Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/JAVA-Developers-Guide/"> JAVA Developer&#39;s Guide</a></li>
          <li><a href="http://www.webbasedprogramming.com/Java-Developers-Reference/" >
           Java Developer&#39;s Reference</a></li>
          <li><a href="http://www.webbasedprogramming.com/Java-Expert-Solutions/" > Java Expert 
          Solutions</a></li>
          <li><a href="http://www.webbasedprogramming.com/Java-Unleashed/" > 
          Java Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/Java-Unleashed-Second-Edition/" > Java 
          Unleashed Second Edition</a></li>
          <li><a href="http://www.webbasedprogramming.com/JavaScript-Manual-of-Style/" > 
          JavaScript Manual of Style</a></li>
          <li><a href="http://www.webbasedprogramming.com/MySQL/" > MySQL</a></li>
          <li><a href="http://www.webbasedprogramming.com/Oracle-Unleashed/" > 
          Oracle Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/Perl-5-By-Example/"> Perl 5 By Example</a></li>
          <li><a href="http://www.webbasedprogramming.com/Perl-5-Unleashed/" >
           Perl 5 Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/Perl-Quick-Reference/" > Perl Quick 
          Reference</a></li>
          <li><a href="http://www.webbasedprogramming.com/PHP/" > PHP</a></li>
          <li><a href="http://www.webbasedprogramming.com/Presenting-JavaBeans/" > Presenting 
          JavaBeans</a></li>
          <li><a href="http://www.webbasedprogramming.com/Special-Edition-Using-Java-Script/" > 
          Special Edition Using Java Script</a></li>
          <li><a href="http://www.webbasedprogramming.com/Special-Edition-Using-Jscript/" > Special 
          Edition Using Jscript</a></li>
          <li><a href="http://www.webbasedprogramming.com/Special-Edition-Using-Visual-FoxPro-6/" > Special 
          Edition Using Visual FoxPro 6</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-CGI-Programming-with-Perl-5-in-a-week-Second-Edit/" > 
          Teach Yourself CGI Programming with Perl 5 in a week Second Edit</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-HTML-3.2-in-24-Hours/" > Teach 
          Yourself HTML 3.2 in 24 Hours</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-Java-1.1-Programming-in-24-Hours/" > 
          Teach Yourself Java 1.1 Programming in 24 Hours</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-Oracle-8-In-21-Days/"> Teach Yourself Oracle 8 In 21 Days</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-Perl-5-in-21-Days/" >
           Teach Yourself Perl 5 in 21 Days</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-SQL-in-21-Days-Second-Edition/" > Teach 
          Yourself SQL in 21 Days Second Edition</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-VBScript-in-21-Days/" > Teach 
          Yourself VBScript in 21 Days</a></li>
          <li><a href="http://www.webbasedprogramming.com/Teach-Yourself-Web-Publishing-with-HTML-3.2-in-14-Days/" > Teach 
          Yourself Web Publishing with HTML 3.2 in 14 Days</a></li>
          <li><a href="http://www.webbasedprogramming.com/The-Complete-Idiots-Guide-to-Creating-an-HTML-Web-Page/" > The 
          Complete Idiot&#39;s Guide to Creating an HTML Web Page</a></li>
          <li><a href="http://www.webbasedprogramming.com/Tricks-of-the-Java-Programming-Gurus/" > Tricks of 
          the Java Programming Gurus</a></li>
          <li><a href="http://www.webbasedprogramming.com/Web-Database-Developers-Guide-with-Visual-Basic-5/" > 
          Web Database Developer&#39;s Guide with Visual Basic 5</a></li>
          <li><a href="http://www.webbasedprogramming.com/Web-Programming-Desktop-Reference-6in1/" > 
          Web Programming Desktop Reference 6in1</a></li>
          <li><a href="http://www.webbasedprogramming.com/Web-Programming-Unleashed/" > Web 
          Programming Unleashed</a></li>
          <li><a href="http://www.webbasedprogramming.com/Web-Programming-with-Perl5/" > Web 
          Programming with Perl5</a></li>
          <li><a href="http://www.webbasedprogramming.com/Web-Based-Systems-and-Network-Management/" > Web-Based 
          Systems and Network Management</a></li>
        </ul>
      </div>



    <div class="lighterBackground">
        <p class="sideBarTitle" align="center"></p>
        <p align="center"><iframe width="120" height="240" src="http://rcm.amazon.com/e/cm?t=uniquecityult-20&o=1&p=8&l=ez&f=ifr" scrolling="no" border="0" frameborder="0" style="border:none;"></iframe>
<br><br></div>

    </div>

    <!-- ##### Main Copy ##### -->

    <div id="main-copy">
  
  <br><center>
  
  
  
  
  
  
  <iframe width="468" height="60" src="http://rcm.amazon.com/e/cm?t=uniquecityult-20&o=1&p=13&l=ez&f=ifr" scrolling="no" border="0" frameborder="0" style="border:none;"></iframe>
  
  
  
  
  
  
  
  
  
  
  
 </center>

  
 
      <h1 id="introduction">Java Unleashed Second Edition</h1>
        <p>
        
        
        
        <HTML>

<HEAD>
   <TITLE>Chapter 10 -- Exception Handling</TITLE>
   <META>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFFFFF" LINK="#0000EE" VLINK="#551A8B" ALINK="#CE2910">
<H1><FONT COLOR=#FF0000>Chapter 10</FONT></H1>
<H1><B><FONT SIZE=5 COLOR=#FF0000>Exception Handling</FONT></B>
</H1>
<P>
<B><I>by David R. Chung</I></B>
<HR WIDTH="100%"></P>
<P>
<H3 ALIGN=CENTER><FONT COLOR="#000000"><FONT SIZE=+2>CONTENTS<A NAME="CONTENTS"></A>
</FONT></FONT></H3>
<UL>
<LI><A HREF="#WhatIsanException" >What Is an Exception?</A>
<LI><A HREF="#IfExceptionsAretheAnswerWhatIsth" >If Exceptions Are the Answer, What Is the Question?</A>
<LI><A HREF="#SomeTerminology" >Some Terminology</A>
<LI><A HREF="#DontThrowUpYourHandsthrowanExcep" >Don't Throw Up Your Hands-throw an Exception</A>
<LI><A HREF="#throwtryandcatchBlocks" >throw, try, and catch Blocks</A>
<LI><A HREF="#TheThrowableClass" >The Throwable Class</A>
<LI><A HREF="#TypesofExceptions" >Types of Exceptions</A>
<LI><A HREF="#BuiltInExceptions" >Built-In Exceptions</A>
<LI><A HREF="#Summary" >Summary</A>
</UL>
<HR>
<P>
Errors are a normal part of programming. Some of these errors
are flaws in a program's basic design or implementation-these
are called <I>bugs</I>. Other types of errors are not really bugs;
rather, they are the result of situations like low memory or invalid
filenames.
<P>
The way you handle the second type of error determines whether
they become bugs. Unfortunately, if your goal is to produce robust
applications, you probably find yourself spending more time handling
errors than actually writing the core of your application.
<P>
Java's exception handling mechanism lets you handle errors without
forcing you to spend most of your energy worrying about them.
<H2><A NAME="WhatIsanException"><FONT SIZE=5 COLOR=#FF0000>What
Is an Exception?</FONT></A></H2>
<P>
As the name implies, an <I>exception</I> is an exceptional condition.
An exception is something that is out of the ordinary. Most often,
exceptions are used as a way to report error conditions. Exceptions
can be used as a means of indicating other situations as well.
This chapter concentrates primarily on exceptions as an error
handling mechanism.
<P>
Exceptions provide notification of errors and a way to handle
them. This new control structure allows you to specify exactly
where to handle specific types of errors.<BR>
<P><CENTER><TABLE BORDER=1 WIDTH=80%>
<TR VALIGN=TOP><TD><B>Note</B></TD></TR>
<TR VALIGN=TOP><TD>
<BLOCKQUOTE>
Other languages such as C++ and Ada provide exception handling. Java's exception handling is similar to the one used by C++.</BLOCKQUOTE>

</TD></TR>
</TABLE></CENTER>
<P>
<H3>Tennyson Understood the Problem</H3>
<P>
In his poem, <I>Charge of the Light Brigade</I>, Alfred, Lord
Tennyson describes an actual battle. In this battle, a cavalry
brigade is ordered to attack a gun emplacement. It turns out that
the valley the troops attack is a trap. There are big guns on
three sides and the brave soldiers on horseback with their sabers
are massacred. The poem describes an actual battle from the Crimean
War.
<P>
The battle as Tennyson describes it highlights a classic problem.
Someone (probably far from the front) had given the order to attack.
The men who led the charge very quickly became aware that an error
had been made. Unfortunately, they did not have the authority
to do anything about it. In Tennyson's immortal words, &quot;Theirs
not to reason why, theirs but to do and die: into the valley of
Death rode the 600.&quot;
<P>
Using exceptions in Java allows you to determine exactly who handles
an error. In fact, low-level functions can detect errors while
higher-level functions decide what to do about them. Exceptions
provide a means of communicating information about errors up through
the chain of methods until one of them can handle it.
<H2><A NAME="IfExceptionsAretheAnswerWhatIsth"><FONT SIZE=5 COLOR=#FF0000>If
Exceptions Are the Answer, What Is the Question?</FONT></A></H2>
<P>
Most procedural languages like C and Pascal do not use exception
handling. In these languages, a variety of techniques are used
to determine whether an error has occurred. The most common means
of error checking is the function's return value.
<P>
Consider the problem of calculating the retail cost of an item
and displaying it. For this example, the retail cost is twice
the wholesale cost:
<BLOCKQUOTE>
<PRE>
int retailCost( int wholesale ) {

    if ( wholesale &lt;= 0 ) {
        return 0 ;
    }

    return (wholesale * 2 ) ;
}
</PRE>
</BLOCKQUOTE>
<P>
The <TT>retailCost()</TT> method takes
the wholesale price of an item and doubles it. If the wholesale
price is negative or zero, the function returns zero to indicate
that an error has occurred. This method can be used in an application
as follows:
<BLOCKQUOTE>
<PRE>
int wholesalePrice = 30 ;
int retailPrice    = 0 ;

retailPrice = retailCost( wholesalePrice ) ;

System.out.println( &quot;Wholesale price = &quot; + wholesalePrice ) ;
System.out.println( &quot;Retail price = &quot;     + retailPrice ) ;
</PRE>
</BLOCKQUOTE>
<P>
In this example, the <TT>retailCost()</TT>
method calculates the correct retail cost and prints it. The problem
is that the code segment never checks whether the <TT>wholesalePrice</TT>
variable is negative. Even though the method checks the value
of <TT>wholesalePrice</TT> and reports
an error-there is nothing that forces the calling method to deal
with the error. If this method is called with a negative <TT>wholesalePrice</TT>,
the function blindly prints invalid data. Therefore, no matter
how diligent you are in ensuring that your methods return error
values, the callers of your methods are free to ignore them.
<P>
You can prevent bad values from being printed by putting the whole
operation in a method. The <TT>showRetail()</TT>
method takes the wholesale price, doubles it, and prints it. If
the wholesale price is negative or zero, the method does not print
anything and returns the <TT>boolean</TT>
value <TT>false</TT>:
<BLOCKQUOTE>
<PRE>
boolean showRetail( int wholesale ) {

    if ( wholesale &lt;= 0 ) {
        return false ;
    }

    int retailPrice ;
    retailPrice = wholesalePrice * 2 ;

    System.out.println( &quot;Wholesale price = &quot; + wholesale ) ;
    System.out.println( &quot;Retail price = &quot;    + retailPrice ) ;
    return true ;

}
</PRE>
</BLOCKQUOTE>
<P>
Using this new and improved method guarantees that <I>bad</I>
values are never printed. However, once again, the caller does
not have to check to see whether the method returned <TT>true</TT>.
<P>
The fact that the caller can choose to ignore return values is
not the only problem with using return values to report errors.
What happens if your method returns a <TT>boolean</TT>
and both <TT>true</TT> and <TT>false</TT>
are valid return values? How does this method report an error?
<P>
Consider a method to determine whether a student passes a test.
The <TT>pass()</TT> method takes the
number of correct answers and the number of questions. The method
calculates the percentage; if it is greater than 70 percent, the
student passes. Consider the <TT>passingGrade()</TT>
method:
<BLOCKQUOTE>
<PRE>
boolean passingGrade( int correct, int total ) {

    boolean returnCode = false ;

    if ( (float)correct / (float)total &gt; 0.70 ) {
        returnCode = true ;
    }

return returnCode ;
}
</PRE>
</BLOCKQUOTE>
<P>
In this example, everything works fine as long as the method arguments
are well behaved. What happens if the number correct is greater
than the total? Or worse, if the total is zero (because this causes
a division by zero in the method)? By relying on return values
in this case, there is <I>no way</I> to report an error in this
function.
<P>
Exceptions prevent you from making your return values do double
duty. Exceptions allow you to use return values to return only
useful information from your methods. Exceptions also <I>force</I>
the caller to deal with errors-because exceptions cannot be ignored.
<H2><A NAME="SomeTerminology"><FONT SIZE=5 COLOR=#FF0000>Some
Terminology</FONT></A></H2>
<P>
Exception handling can be viewed as a nonlocal control structure.
When a method <I>throws</I> an exception, its caller must determine
whether it can <I>catch</I> the exception. If the calling method
can catch the exception, it takes over and execution continues
in the caller. If the calling method cannot catch the exception,
the exception is passed on to its caller. This process continues
until either the exception is caught or the top (or bottom, depending
on how you look at it) of the call stack is reached and the application
terminates because the exception has not been caught.
<P>
Java exceptions are class objects subclassed from <TT>java.lang.Throwable</TT>.
Because exceptions are class objects, they can contain both data
and methods. In fact, the base class <TT>Throwable</TT>
implements a method that returns a <TT>String</TT>
describing the error that caused the exception. This is useful
for debugging and for giving users a meaningful error message.
<H2><A NAME="DontThrowUpYourHandsthrowanExcep"><FONT SIZE=5 COLOR=#FF0000>Don't
Throw Up Your Hands-</FONT><FONT COLOR=#FF0000><TT>throw</FONT></TT><FONT SIZE=5 COLOR=#FF0000>
an Exception</FONT></A></H2>
<P>
The <TT>passingGrade()</TT> method
presented in the preceding section was unable to report an error
condition because all its possible return values were valid. Adding
exception handling to the method makes it possible to uncouple
the reporting of results from the reporting of errors.
<P>
The first step is to modify the <TT>passingGrade()</TT>
method definition to include the <TT>throws</TT>
clause. The <TT>throws</TT> clause
lists the types of exceptions that can be thrown by the method.
In the following revised code, the method throws only an exception
of type <TT>Exception</TT>:
<BLOCKQUOTE>
<PRE>
static boolean passingGrade( int correct, int total )
                                    throws Exception {

    boolean returnCode = false ;
</PRE>
</BLOCKQUOTE>
<P>
The rest of the method remains largely unchanged. This time, the
method checks to see whether its arguments make sense. Because
this method determines passing grades, it would be un-reasonable
to have more correct responses than total responses. Therefore,
if there are more correct responses than total responses, the
method throws an exception.
<P>
The method instantiates an object of type <TT>Exception</TT>.
The <TT>Exception</TT> constructor
takes a <TT>String</TT> parameter.
The <TT>String</TT> contains a message
that can be retrieved when the exception is caught. The <TT>throw</TT>
statement terminates the method and gives its caller the opportunity
to catch it:
<BLOCKQUOTE>
<PRE>
    if( correct &gt; total ) {
        throw new Exception( &quot;Invalid values&quot; ) ;
    }

    if ( (float)correct / (float)total &gt; 0.70 ) {
        returnCode = true ;
    }

    return returnCode ;
}
</PRE>
</BLOCKQUOTE>
<H2><A NAME="throwtryandcatchBlocks"><TT><FONT COLOR=#FF0000>throw</FONT></TT><FONT SIZE=5 COLOR=#FF0000>,
</FONT><TT><FONT COLOR=#FF0000>try</FONT></TT><FONT SIZE=5 COLOR=#FF0000>,
and </FONT><TT><FONT COLOR=#FF0000>catch</TT></FONT><FONT SIZE=5 COLOR=#FF0000>
Blocks</FONT></A></H2>
<P>
To respond to an exception, the call to the method that produces
it must be placed within a <TT>try</TT>
block. A <TT>try</TT> block is a block
of code beginning with the <TT>try</TT>
keyword followed by a left and a right curly brace. Every <TT>try</TT>
block is associated with one or more <TT>catch</TT>
blocks. Here is a <TT>try</TT> block:
<BLOCKQUOTE>
<PRE>
try
    {
    // method calls go here
    }
</PRE>
</BLOCKQUOTE>
<P>
If a method is to catch exceptions thrown by the methods it calls,
the calls must be placed within a <TT>try</TT>
block. If an exception is thrown, it is handled in a <TT>catch</TT>
block. Different <TT>catch</TT> blocks
handle different types of exceptions. This is a <TT>try</TT>
block and a <TT>catch</TT> block set
up to handle exceptions of type <TT>Exception</TT>:
<BLOCKQUOTE>
<PRE>
try
    {
    // method calls go here
    }
catch( Exception e )
    {
    // handle exceptons here
    }
</PRE>
</BLOCKQUOTE>
<P>
When any method in the <TT>try</TT>
block throws any type of exception, execution of the <TT>try</TT>
block ceases. Program control passes immediately to the associated
<TT>catch</TT> block. If the <TT>catch</TT>
block can handle the given exception type, it takes over. If it
cannot handle the exception, the exception is passed to the method's
caller. In an application, this process goes on until a <TT>catch</TT>
block catches the exception or the exception reaches the <TT>main()</TT>
method uncaught and causes the application to terminate.
<H3>An Exceptional Example</H3>
<P>
Because <I>all</I> Java methods are class members, the <TT>passingGrade()</TT>
method is incorporated in the <TT>gradeTest</TT>
application class. Because <TT>main()</TT>
calls <TT>passingGrade()</TT>, <TT>main()</TT>
must be able to catch any exceptions <TT>passingGrade()</TT>
might throw. To do this, <TT>main()</TT>
places the call to <TT>passingGrade()</TT>
in a <TT>try</TT> block. Because the
<TT>throws</TT> clause lists type
<TT>Exception</TT>, the <TT>catch</TT>
block catches the <TT>Exception</TT>
class. Listing 10.1 shows the entire <TT>gradeTest</TT>
application.
<HR>
<P>
<B>Listing 10.1. The </B><TT><B>gradeTest</B></TT><B>
application.<BR>
</B>
<BLOCKQUOTE>
<PRE>
import java.io.* ;
import java.lang.Exception ;

public class gradeTest {

    public static void main( String[] args ) {

        try
            {
            // the second call to passingGrade throws
            // an excption so the third call never
            // gets executed

            System.out.println( passingGrade( 60,  80 ) ) ;
            System.out.println( passingGrade( 75,   0 ) ) ;
            System.out.println( passingGrade( 90, 100 ) ) ;
            }
        catch( Exception e )
            {
            System.out.println( &quot;Caught exception --&quot; +
                                e.getMessage() ) ;
            }

    }

    static boolean passingGrade( int correct, int total )
                                        throws Exception {

        boolean returnCode = false ;

        if( correct &gt; total ) {
            throw new Exception( &quot;Invalid values&quot; ) ;
        }

        if ( (float)correct / (float)total &gt; 0.70 ) {
            returnCode = true ;
        }

        return returnCode ;
    }

}
</PRE>
</BLOCKQUOTE>
<HR>
<P>
The second call to <TT>passingGrade()</TT>
fails in this case, because the method checks to see whether the
number of correct responses is less than the total responses.
When <TT>passingGrade()</TT> throws
the exception, control passes to the <TT>main()</TT>
method. In this example, the <TT>catch</TT>
block in <TT>main()</TT> catches the
exception and prints <TT>Caught exception
-- Invalid values</TT>.
<H3>Multiple <TT>catch</TT><FONT SIZE=4>
Blocks</FONT></H3>
<P>
In some cases, a method may have to catch different types of exceptions.
Java supports multiple <TT>catch</TT>
blocks. Each <TT>catch</TT> block
must specify a different type of exception:
<BLOCKQUOTE>
<PRE>
try
    {
    // method calls go here
    }
catch( SomeExceptionClass e )
    {
    // handle SomeExceptionClass exceptions here
    }
catch( SomeOtherExceptionClass e )
    {
    // handle SomeOtherExceptionClass exceptions here
    }
</PRE>
</BLOCKQUOTE>
<P>
When an exception is thrown in the <TT>try</TT>
block, it is caught by the first <TT>catch</TT>
block of the appropriate type. Only one <TT>catch</TT>
block in a given set will be executed. Notice that the <TT>catch</TT>
block looks a lot like a method declaration. The exception caught
in a <TT>catch</TT> block is a local
reference to the actual exception object. You can use this exception
object to help determine what caused the exception to be thrown
in the first place.
<H3>Does Every Method Have to Catch Every Exception?</H3>
<P>
What happens if a method calls another method that throws an exception
but chooses not to catch it? In the example in Listing 10.2, <TT>main()</TT>
calls <TT>foo()</TT>, which in turn
calls <TT>bar()</TT>. <TT>bar()</TT>
lists <TT>Exception</TT> in its <TT>throws</TT>
clause; because <TT>foo()</TT> is
not going to catch the exception, it must also have <TT>Exception</TT>
in its <TT>throws</TT> clause. The
application in Listing 10.2 shows a method, <TT>foo(),</TT>
that ignores exceptions thrown by the called method.
<HR>
<P>
<B>Listing 10.2. A method that ignores exceptions thrown by the
method it calls.<BR>
</B>
<BLOCKQUOTE>
<PRE>
import java.io.* ;
import java.lang.Exception ;

public class MultiThrow {

    public static void main( String[] args ) {

        try
            {
            foo() ;
            }
        catch( Exception e )
            {
            System.out.println( &quot;Caught exception &quot; +
                                e.getMessage() ) ;
            }

    }

    static void foo() throws Exception {

        bar() ;

    }

    static void bar() throws Exception {

        throw new Exception( &quot;Who cares&quot; ) ;

    }

}
</PRE>
</BLOCKQUOTE>
<HR>
<P>
In the example in Listing 10.3, <TT>main()</TT>
calls <TT>foo()</TT> which calls <TT>bar()</TT>.
Because <TT>bar()</TT> throws an exception
and doesn't catch it, <TT>foo()</TT>
has the opportunity to catch it. The <TT>foo()</TT>
method has no <TT>catch</TT> block,
so it cannot catch the exception. In this case, the exception
propagates up the call stack to <TT>foo()</TT>'s
caller, <TT>main()</TT>.
<HR>
<P>
<B>Listing 10.3. A method that catches and rethrows an exception.
<BR>
</B>
<BLOCKQUOTE>
<PRE>
import java.io.* ;
import java.lang.Exception ;
public class MultiThrow {

    public static void main( String[] args ) {

        try
            {
            foo() ;
            }
        catch( Exception e )
            {
            System.out.println( &quot;Caught exception &quot; +
                                e.getMessage() ) ;
            }

    }

    static void foo() throws Exception {

        try
            {
            bar() ;
            }
        catch( Exception e )
            {
            System.out.println( &quot;Re throw exception -- &quot; +
                                e.getMessage() ) ;
            throw e ;
            }    }

    static void bar() throws Exception {

        throw new Exception( &quot;Who cares&quot; ) ;

    }

}
</PRE>
</BLOCKQUOTE>
<HR>
<P>
The <TT>foo()</TT> method calls <TT>bar()</TT>.
The <TT>bar()</TT> method throws an
exception and <TT>foo()</TT> catches
it. In this example, <TT>foo()</TT>
simply <I>rethrows</I> the exception, which is ultimately caught
in the application's <TT>main()</TT>
method. In a <I>real</I> application, <TT>foo()</TT>
could do some processing and then rethrow the exception. This
arrangement allows both <TT>foo()</TT>
and <TT>main()</TT> to handle the
exception.
<H3>The <TT>finally</TT><FONT SIZE=4>
Clause</FONT></H3>
<P>
Java introduces a new concept in exception handling: the <TT>finally</TT>
clause. The <TT>finally</TT> clause
sets apart a block of code that is always executed. Here's an
example of a <TT>finally</TT> clause:
<BLOCKQUOTE>
<PRE>
import java.io.* ;
import java.lang.Exception ;
public class MultiThrow {
    public static void main( String[] args ) {

        try
            {
            alpha() ;
            }
       catch( Exception e }
            {
            System.out.println( &quot;Caught exception &quot; ) ;
            }
       finally()
            {
            System.out.println( &quot;Finally. &quot; ) ;
            }
    }

}
</PRE>
</BLOCKQUOTE>
<P>
In normal execution (that is, when no exceptions are thrown),
the <TT>finally</TT> block is executed
immediately after the <TT>try</TT>
block. When an exception is thrown, the <TT>finally</TT>
block is executed before control passes to the caller.
<P>
If <TT>alpha()</TT> throws an exception,
it is caught in the <TT>catch</TT>
block and <I>then</I> the <TT>finally</TT>
block is executed. If <TT>alpha()</TT>
does not throw an exception, the <TT>finally</TT>
block is executed after the <TT>try</TT>
block. If any code in a <TT>try</TT>
block is executed, the <TT>finally</TT>
block is executed as well.
<H2><A NAME="TheThrowableClass"><FONT SIZE=5 COLOR=#FF0000>The
</FONT><TT><FONT COLOR=#FF0000>Throwable</FONT></TT><FONT SIZE=5 COLOR=#FF0000>
Class</FONT></A></H2>
<P>
All exceptions in Java are subclassed from the class <TT>Throwable</TT>.
If you want to create your own exception classes, you must subclass
<TT>Throwable</TT>. Most Java programs
do not have to subclass their own exception classes.
<P>
Following is the <TT>public</TT> portion
of the class definition of <TT>Throwable</TT>:
<BLOCKQUOTE>
<PRE>
public class Throwable {

    public Throwable() ;
    public Throwable(String message) ;
    public String getMessage()
    public String toString() ;
    public void printStackTrace() ;
    public void printStackTrace(java.io.PrintStream s) ;
    private native void printStackTrace0(java.io.PrintStream s);
    public native Throwable fillInStackTrace();

}
</PRE>
</BLOCKQUOTE>
<P>
The constructor takes a string that describes the exception. Later,
when an exception is thrown, you can call the <TT>getMessage()</TT>
method to get the error string that was reported.
<H2><A NAME="TypesofExceptions"><FONT SIZE=5 COLOR=#FF0000>Types
of Exceptions</FONT></A></H2>
<P>
The methods of the Java API and the language itself also throw
exceptions. These exceptions can be broken into two classes: <TT>Exception
</TT>and <TT>Error</TT>.
<P>
Both the <TT>Exception</TT> and <TT>Error</TT>
classes are derived from <TT>Throwable</TT>.
<TT>Exception</TT> and its subclasses
are used to indicate conditions that may be recoverable. <TT>Error</TT>
and its subclasses indicate conditions that are generally not
recoverable and should cause your applet to terminate.
<P>
The various packages included in the Java Developers Kit throw
different kinds of <TT>Exception</TT>
and <TT>Error</TT> exceptions, as
described in the following sections.
<H3><TT>java.lang</TT><FONT SIZE=4>
Exceptions</FONT></H3>
<P>
The <TT>java.lang</TT> package contains
much of the core Java language. The exceptions subclassed from
<TT>RuntimeException</TT> do not have
to be declared in a method's <TT>throws</TT>
clause. These exceptions are considered <I>normal</I> and nearly
any method can throw them. Figure 10.1 and Table 10.1 show the
recoverable exceptions from the <TT>java.lang</TT>
package. Figure 10.2 and Table 10.2 show the nonrecoverable errors
in the <TT>java.lang</TT> package.
<P>
<A HREF="f10-1.gif" ><B>Figure 10.1:</B> <I>The </I>Java.lang <I>exception hierarchy.</I></A>
<P>
<A HREF="f10-2.gif" ><B>Figure 10.2:</B> <I>The</I> Java.lang <I>error hierarchy.</I></A>
<BR>
<P>
<CENTER><B>Table 10.1. The </B><TT><B>java.lang</B></TT><B>
exceptions.</B></CENTER>
<P><CENTER><TABLE BORDER=1 WIDTH=80%>
<TR VALIGN=TOP><TD WIDTH=259><I>Exception</I></TD><TD WIDTH=331><I>Cause</I>
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>ArithmeticException </TT>
</TD><TD WIDTH=331>Arithmetic error condition (for example, divide by zero).
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>ArrayIndexOutOfBoundsException</TT> 
</TD><TD WIDTH=331>Array index is less than zero or greater than the actual size of the array.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>ArrayStoreException</TT> 
</TD><TD WIDTH=331>Object type mismatch between the array and the object to be stored in the array.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>ClassCastException</TT> 
</TD><TD WIDTH=331>Cast of object to inappropriate type.</TD>
</TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>ClassNotFoundException</TT> 
</TD><TD WIDTH=331>Unable to load the requested class.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>CloneNotSupportedException</TT>
</TD><TD WIDTH=331>Object does not implement the <TT>cloneable</TT> interface.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>Exception</TT> 
</TD><TD WIDTH=331>Root class of the exception hierarchy.</TD>
</TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>IllegalAccessException</TT> 
</TD><TD WIDTH=331>Class is not accessible.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>IllegalArgumentException</TT>
</TD><TD WIDTH=331>Method receives an illegal argument.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>IllegalMonitorStateException</TT>
</TD><TD WIDTH=331>Improper monitor state (thread synchronization).
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>IllegalThreadStateException</TT>
</TD><TD WIDTH=331>The thread is in an improper state for the requested operation.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>IndexOutOfBoundsException</TT> 
</TD><TD WIDTH=331>Index is out of bounds.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>InstantiationException</TT> 
</TD><TD WIDTH=331>Attempt to create an instance of the abstract class.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>InterruptedException</TT> 
</TD><TD WIDTH=331>Thread interrupted.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>NegativeArraySizeException</TT> 
</TD><TD WIDTH=331>Array size is less than zero.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>NoSuchMethodException</TT> 
</TD><TD WIDTH=331>Unable to resolve method.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>NullPointerException</TT> 
</TD><TD WIDTH=331>Attempt to access a <TT>null</TT> object member.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>NumberFormatException</TT> 
</TD><TD WIDTH=331>Unable to convert the string to a number.</TD>
</TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>RuntimeException</TT> 
</TD><TD WIDTH=331>Base class for many <TT>java.lang</TT> exceptions.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>SecurityException</TT> 
</TD><TD WIDTH=331>Security settings do not allow the operation.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=259><TT>StringIndexOutOfBoundsException</TT> 
</TD><TD WIDTH=331>Index is negative or greater than the size of the string.
</TD></TR>
</TABLE></CENTER>
<P>
<P>
<CENTER><B>Table 10.2. The </B><TT><B>java.lang</B></TT><B>
errors.</B></CENTER>
<P><CENTER><TABLE BORDER=1 WIDTH=80%>
<TR VALIGN=TOP><TD WIDTH=295><I>Error</I></TD><TD WIDTH=295><I>Cause</I>
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>AbstractMethodError</TT>
</TD><TD WIDTH=295>Attempt to call an abstract method.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>ClassCircularityError</TT>
</TD><TD WIDTH=295>This error is no longer used.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>ClassFormatError </TT>
</TD><TD WIDTH=295>Invalid binary class format.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>Error </TT>
</TD><TD WIDTH=295>Root class of the error hierarchy.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>IllegalAccessError </TT>
</TD><TD WIDTH=295>Attempt to access an inaccessible object.</TD>
</TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>IncompatibleClassChangeError</TT> 
</TD><TD WIDTH=295>Improper use of a class.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>InstantiationError </TT>
</TD><TD WIDTH=295>Attempt to instantiate an abstract class.</TD>
</TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>InternalError </TT>
</TD><TD WIDTH=295>Error in the interpreter.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>LinkageError </TT>
</TD><TD WIDTH=295>Error in class dependencies.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>NoClassDefFoundError </TT>
</TD><TD WIDTH=295>Unable to find the class definition.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>NoSuchFieldError </TT>
</TD><TD WIDTH=295>Unable to find the requested field.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>NoSuchMethodError </TT>
</TD><TD WIDTH=295>Unable to find the requested method.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>OutOfMemoryError </TT>
</TD><TD WIDTH=295>Out of memory.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>StackOverflowError </TT>
</TD><TD WIDTH=295>Stack overflow.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>ThreadDeath </TT>
</TD><TD WIDTH=295>Indicates that the thread will terminate. May be caught to perform cleanup. (If caught, must be rethrown.)
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>UnknownError </TT>
</TD><TD WIDTH=295>Unknown virtual machine error.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>UnsatisfiedLinkError </TT>
</TD><TD WIDTH=295>Unresolved links in the loaded class.</TD>
</TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>VerifyError </TT>
</TD><TD WIDTH=295>Unable to verify bytecode.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=295><TT>VirtualMachineError</TT>
</TD><TD WIDTH=295>Root class for virtual machine errors.</TD>
</TR>
</TABLE></CENTER>
<P>
<H3><TT>java.io</TT><FONT SIZE=4>
Exceptions</FONT></H3>
<P>
The classes in <TT>java.io</TT> throw
a variety of exceptions, as shown in Table 10.3 and Figure 10.3.
Any classes that work with I/O are good candidates to throw recoverable
exceptions. For example, activities such as opening files or writing
to files are likely to fail from time to time. The classes of
the <TT>java.io</TT> package do not
throw errors at all.
<P>
<A HREF="f10-3.gif" ><B>Figure 10.3:</B> <I>The</I> Java.io exception hierarchy.</A>
<BR>
<P>
<CENTER><B>Table 10.3. The </B><TT><B>java.io</B></TT><B>
exceptions.</B></CENTER>
<P><CENTER><TABLE BORDER=1 WIDTH=80%>
<TR VALIGN=TOP><TD WIDTH=234><I>Exception</I></TD><TD WIDTH=356><I>Cause</I>
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=234><TT>IOException</TT>
</TD><TD WIDTH=356>Root class for I/O exceptions.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=234><TT>EOFException</TT>
</TD><TD WIDTH=356>End of file.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=234><TT>FileNotFoundException </TT>
</TD><TD WIDTH=356>Unable to locate the file.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=234><TT>InterruptedIOException</TT>
</TD><TD WIDTH=356>I/O operation was interrupted. Contains a <TT>bytesTransferred</TT> member that indicates how many bytes were transferred before the operation was interrupted.
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=234><TT>UTFDataFormatException</TT>
</TD><TD WIDTH=356>Malformed UTF-8 string.</TD></TR>
</TABLE></CENTER>
<P>
<H3><TT>java.net</TT><FONT SIZE=4>
Exceptions</FONT></H3>
<P>
The <TT>java.net</TT> package handles
network communications. Its classes most often throw exceptions
to indicate connect failures and the like. Table 10.4 and Figure
10.4 show the recoverable exceptions from the <TT>java.net</TT>
package. The classes of the <TT>java.net</TT>
package do not throw errors at all.
<P>
<A HREF="f10-4.gif" ><B>Figure 10.4:</B> <I>The </I>Java.net <I>exception hirearchy.</I></A>
<BR>
<P>
<CENTER><B>Table 10.4. The </B><TT><B>java.net</B></TT><B>
exceptions.</B></CENTER>
<P><CENTER><TABLE BORDER=1 WIDTH=80%>
<TR VALIGN=TOP><TD WIDTH=244><I>Exception</I></TD><TD WIDTH=281><I>Cause</I>
</TD></TR>
<TR VALIGN=TOP><TD WIDTH=244><TT>MalformedURLException</TT>
</TD><TD WIDTH=281>Unable to interpret URL.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=244><TT>ProtocolException</TT>
</TD><TD WIDTH=281>Socket class protocol error.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=244><TT>SocketException</TT>
</TD><TD WIDTH=281>Socket class exception.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=244><TT>UnknownHostException</TT>
</TD><TD WIDTH=281>Unable to resolve the host name.</TD></TR>
<TR VALIGN=TOP><TD WIDTH=244><TT>UnknownServiceException</TT>
</TD><TD WIDTH=281>Connection does not support the service.</TD>
</TR>
</TABLE></CENTER>
<P>
<H3><TT>java.awt</TT><FONT SIZE=4>
Exceptions</FONT></H3>
<P>
The AWT classes have members that throw one error and one exception:
<UL>
<LI><TT>AWTException</TT> (exception
in AWT)
<LI><TT>AWTError</TT> (error in AWT)
</UL>
<H3><TT>java.util</TT><FONT SIZE=4>
Exceptions</FONT></H3>
<P>
The classes of <TT>java.util</TT>
throw the following exceptions:
<UL>
<LI><TT>EmptyStackException</TT> 
(no objects on stack)
<LI><TT>NoSuchElementException</TT>
(no more objects in collection)
</UL>
<H2><A NAME="BuiltInExceptions"><FONT SIZE=5 COLOR=#FF0000>Built-In
Exceptions</FONT></A></H2>
<P>
In the example in Listing 10.4, you see how the <I>automatic</I>
exceptions in Java work. This application creates a method and
forces it to divide by zero. The method does not have to explicitly
throw an exception because the division operator throws an exception
when required.
<HR>
<P>
<B>Listing 10.4. An example of a built-in exception.<BR>
</B>
<BLOCKQUOTE>
<PRE>
import java.io.* ;
import java.lang.Exception ;

public class DivideBy0 {

    public static void main( String[] args ) {

    int a = 2 ;
    int b = 3 ;
    int c = 5 ;
    int d = 0 ;
    int e = 1 ;
    int f = 3 ;

    try
        {
        System.out.println( a+&quot;/&quot;+b+&quot; = &quot;+div( a, b ) ) ;
        System.out.println( c+&quot;/&quot;+d+&quot; = &quot;+div( c, d ) ) ;
        System.out.println( e+&quot;/&quot;+f+&quot; = &quot;+div( e, f ) ) ;
        }
    catch( Exception except )
        {
        System.out.println( &quot;Caught exception &quot; +
                                except.getMessage() ) ;
        }
    }

    static int div( int a, int b ) {

        return (a/b) ;

    }

}
The output of this application is shown here:
2/3 = 0
Caught exception / by zero
</PRE>
</BLOCKQUOTE>
<HR>
<P>
The first call to <TT>div()</TT> works
fine. The second call fails because of the divide-by-zero error.
Even though the application did not specify it, an exception was
thrown-and caught. So you can use arithmetic in your code without
writing code that explicitly checks bounds.
<H2><A NAME="Summary"><FONT SIZE=5 COLOR=#FF0000>Summary</FONT></A>
</H2>
<P>
The exception handling mechanism in Java allows your methods to
report errors in a manner that cannot be ignored. Every exception
that is thrown must be caught or the application terminates. Exceptions
are actually class objects derived from the <TT>Throwable</TT>
class. Therefore, exceptions combine data and methods; an exception
object generally contains a string explaining what the error is.
<P>
Exception handling helps you combine error processing in one place.
It uncouples the reporting of results and the reporting of errors.
If you use exception handling, you can create much more powerful
and robust code.
<P>
<HR WIDTH="100%"></P>
<CENTER><P><A HREF="ch9.htm"><IMG SRC="pc.gif" BORDER=0 HEIGHT=88 WIDTH=140></A>
<A HREF="#CONTENTS"><IMG SRC="cc.gif" BORDER=0 HEIGHT=88 WIDTH=140></A>
<A HREF="contents.htm"><IMG SRC="hb.gif" BORDER=0 HEIGHT=88 WIDTH=140></A>
<A HREF="ch11.htm"><IMG SRC="nc.gif" BORDER=0 HEIGHT=88 WIDTH=140></A></P></CENTER>
<P>
<HR WIDTH="100%"></P>
</BODY>
</HTML>


        
        
        </p>

    </div>
    
    <!-- ##### Footer ##### -->

    <div id="footer">
      <div class="left">
        Contact Webmaster:&nbsp;<a href="mailto:webmaster@webbasedprogramming.com">webmaster@webbasedprogramming.com</a></div>

      <br class="doNotDisplay doNotPrint" />

      <div class="right">
        All content is © to respective authors.<br />
        &nbsp;</div>
    </div>
  </body>
</html>
